contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,826 @@
+as follows.
+    1. If the Pod's ordinal is in the sequence `[0,.Status.CurrentReplicas)`, 
+    the Pod should be consistent with version indicated by `Status.CurrentRevision`.
+    1. If the Pod's ordinal is in the sequence 
+    `[.Status.Replicas - .Status.UpdatedReplicas, .Status.Replicas)`
+    the Pod should be consistent with the version indicated by 
+    `Status.UpdateRevision`.
+1. If the StatefulSet's `.Spec.UpdateStrategy.Type` is equal to 
+`PartitionStatefulSetStrategyType` then the version of the Pod should be 
+as follows.
+    1. If the Pod's ordinal is in the sequence `[0,.Status.CurrentReplicas)`, 
+    the Pod should be consistent with version indicated by `Status.CurrentRevision`.
+    1. If the Pod's ordinal is in the sequence 
+    `[.Status.Replicas - .Status.UpdatedReplicas, .Status.Replicas)`the Pod 
+    should be consistent with the version indicated by `Status.UpdateRevision`.
+    1. If the Pod does not meet either of the prior two conditions, and if 
+    ordinal is in the sequence `[0, .Spec.UpdateStrategy.Partition.Ordinal)`, 
+    it should  be consistent with the version indicated by 
+    `Status.CurrentRevision`.
+    1. Otherwise, the Pod should be consistent with the version indicated 
+    by `Status.UpdateRevision`.
+
+### Pod State Reconciliation
+In order to reconcile a Pod with declared desired 
+[target state](#target-pod-state) the StatefulSet controller will do the 
+following.
+
+1. If the Pod is already consistent with its target state the controller will do 
+nothing.
+1. If the Pod is labeled with a `StatefulSetRevisionLabel` that indicates 
+the Pod was generated from a version of the StatefulSetSpec that is semantically 
+equivalent to, but not equal to, the [target version](#target-pod-state), the 
+StatefulSet controller will update the Pod with a `StatefulSetRevisionLabel` 
+indicating the new semantically equivalent version. This form of reconciliation 
+is non-destructive.
+1. If the Pod was not created from the target version, the Pod will be deleted 
+and recreated from that version. This form of reconciliation is destructive.
+
+### Specification Updates
+The StatefulSet controller will [snapshot](#snapshot-creation) its target 
+Object state when mutations are made to its `.Spec.Template` or 
+`.Spec.VolumeClaimsTemplate` (Note that the latter is currently immutable).
+
+1. When the StatefulSet controller observes a mutation to a StatefulSet's 
+ `.Spec.Template` it will snapshot its target Object state and compare 
+the snapshot with the version indicated by its `.Status.UpdateRevision`.
+1. If the current state is equivalent to the version indicated by 
+`.Status.UpdateRevision` no update has occurred. 
+1. If the `Status.CurrentRevision` field is empty, then the StatefulSet has no 
+revision history. To initialize its revision history, the StatefulSet controller 
+will set both `.Status.CurrentRevision` and `.Status.UpdateRevision` to the 
```

Comment:
When the StatefulSet controller initializes its history via the CurrentRevision and UpdateRevision it will have to reconcile the target state of each Pod. If the history is initialized due to creation of a new StatefulSet and no Pods exist, all Pods will be created with CurrentRevsion and be CurrentReplicas. If the update is due to a mutation to an existing set that causes its history to be initialized, the controller will have to consider the revision that each Pod is labeled with and reconcile its state appropriately.  In the case where no Pods are labled and the CurrentRevision and UpdateRevision are equivalent, all Pods need to be deleted and recreated from this revision.